### PR TITLE
8278479: RunThese test failure with +UseHeavyMonitors and +VerifyHeavyMonitors

### PR DESCRIPTION
**Pull request by coleenp**
This change adds a null check before calling into Runtime1::monitorenter when -XX:+UseHeavyMonitors is set. There's a null check in the C2 and interpreter code before calling the runtime function but not C1.

Tested with tier1-7 (a little of 8) and built on most non-oracle platforms as well.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8278479](https://bugs.openjdk.org/browse/JDK-8278479): RunThese test failure with +UseHeavyMonitors and +VerifyHeavyMonitors

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [912dc5cd](https://git.openjdk.org/jdk/pull/9339/files/912dc5cdfc7ae9678b9bc7dc86043699f3d335f3)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to [912dc5cd](https://git.openjdk.org/jdk/pull/9339/files/912dc5cdfc7ae9678b9bc7dc86043699f3d335f3)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [912dc5cd](https://git.openjdk.org/jdk/pull/9339/files/912dc5cdfc7ae9678b9bc7dc86043699f3d335f3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9339/head:pull/9339` \
`$ git checkout pull/9339`

Update a local copy of the PR: \
`$ git checkout pull/9339` \
`$ git pull https://git.openjdk.org/jdk pull/9339/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9339`

View PR using the GUI difftool: \
`$ git pr show -t 9339`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9339.diff">https://git.openjdk.org/jdk/pull/9339.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by coleenp** [comment]
Thanks Vladimir and Dan for the quick reviews.
I ran all of the tiers with UseHeavyMonitors on, except I stopped the long running tests in tier8 because they would have failed right away if it was wrong.
The test for NullPointerExceptionMessage directly exercises this code with the options.
RunThese didn't reproduce this bug locally or with several non-local runs but the stack was the same as the NPE test.

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp [hunk 1]
```diff
@@ -2558,6 +2558,11 @@ void LIR_Assembler::emit_lock(LIR_OpLock* op) {
   Register hdr = op->hdr_opr()->as_register();
   Register lock = op->lock_opr()->as_register();
   if (UseHeavyMonitors) {
+    if (op->info() != NULL) {
+      int null_check_offset = __ offset();
+      __ null_check(obj, -1);
```
Comment (dean-long) [inline src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:2563]:
```suggestion
      __ null_check(obj);
```

File: src/hotspot/cpu/arm/c1_LIRAssembler_arm.cpp [hunk 2]
```diff
@@ -2432,6 +2432,11 @@ void LIR_Assembler::emit_lock(LIR_OpLock* op) {
   Register lock = op->lock_opr()->as_pointer_register();
 
   if (UseHeavyMonitors) {
+    if (op->info() != NULL) {
+      int null_check_offset = __ offset();
+      __ null_check(obj);
+      add_debug_info_for_null_check(null_check_offset, op->info());
```
Comment (dean-long) [inline src/hotspot/cpu/arm/c1_LIRAssembler_arm.cpp:2438]: Is this equivalent to the following?
```
add_debug_info_for_null_check_here(op->info());
__ null_check(obj);
```

**Review by dean-long — APPROVED** [review]

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:2563]: Yes, that would be better but it leads to a compilation error on macosx-aarch64 that null_check is ambiguous:

src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:2563:10: error: call to member function 'null_check' is ambiguous
[2022-06-30T00:09:19,094Z]       __ null_check(obj);
[2022-06-30T00:09:19,094Z]       ~~~^~~~~~~~~~
src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:602:16: note: candidate function
[2022-06-30T00:09:19,094Z]   virtual void null_check(Register reg, int offset = -1);
[2022-06-30T00:09:19,094Z]                ^
src/hotspot/cpu/aarch64/c1_MacroAssembler_aarch64.hpp:109:8: note: candidate function
[2022-06-30T00:09:19,094Z]   void null_check(Register r, Label *Lnull = NULL) { MacroAssembler::null_check(r); }
[2022-06-30T00:09:19,094Z]        ^

File: src/hotspot/cpu/arm/c1_LIRAssembler_arm.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/cpu/arm/c1_LIRAssembler_arm.cpp:2438]: I don't know.  Is it better and preferable?

Comment (dean-long) [reply src/hotspot/cpu/arm/c1_LIRAssembler_arm.cpp]: I like it better.  Thanks for changing it.

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp (hunk 1 above)
Comment (dean-long) [reply src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:2563]: OK, nevermind.

**Comment by coleenp** [comment]
Thanks Dean for the review and improvement.
/integrate

**Pull request closed without merging**